### PR TITLE
Add support for remediation commits

### PR DESCRIPTION
**Pull request by brianwarner**
This PR adds support for remediation commits and expands test coverage. In brief, a remediation commit is an additional way for projects to recognize valid, retroactive signoffs on commits which otherwise fail the Probot DCO checks. 

Remediations come in two flavors. Individual remediation commits allow an author to fix one or more failing commits by pushing an additional, properly signed-off commit with specific text indicating they apply their signoff retroactively. Third-party remediations allow someone other than the author to do this.

**About this PR**

First and foremost the default Probot DCO behavior is unchanged, although the default remediation text and recommendations are modified (for reasons explained in the commit message). Unless a project explicitly enables remediation commits via their repo's `.github/dco.yml` file, it will continue to strictly enforce the DCO and only permit commits that explicitly include a proper Signed-off-by line.

The first commit ([333a52f](https://github.com/probot/dco/commit/333a52f472923773519e3d8db00da552e1458d67)) expands and normalizes the default test cases. Running `./node_modules/jest/bin/jest.js --verbose` now returns a structured view of what we're testing, organized into success and failure scenarios. This helps ensure the default behavior remains constant through the subsequent changes.

The rest of the commits incrementally add support for remediation commits. [5aecad9](https://github.com/probot/dco/commit/5aecad9289eb135a39748306a85de032c92bc5db) modifies and consolidates the function that explains how to fix a failing PR, because it will be important later on. I also found a corner case in testing that makes me think it's better to just recommend all default fixes be rebases instead of amends. Basically if you push multiple commits and only one fails, and it's not the most recent commit, `git commit --amend` is going to fix the most recent commit and isn't going to fix the issue. Since both approaches required a force-push anyhow, it seemed less error prone to just recommend a rebase.

Next, [1a43157](https://github.com/probot/dco/commit/1a4315745dceeba8a15b3ec3554b14efb6d1e42e) adds full support for remediation commits. When a commit fails, the user gets specific copy-and-paste code blocks they can use to fix the problem. In addition the test cases are increased by 4x, because in addition to testing new functionality, they also repeat all of the previous tests to ensure enabling the config options does not mess with default results.

Finally [0696f05](https://github.com/probot/dco/commit/0696f05c1bef5690d81fa7ecab08116743996df2) updates the README.

**Why add this feature**

At the Linux Foundation we've heard feedback from some of our communities that DCO failures are trivial for some developers to remediate, but a barrier for others. In particular, some projects have expressed concern that one-time contributors are opening PRs through the GitHub web UI, are getting blocked because they don't know to add a signoff, and are simply abandoning their contributions rather than set up a local git environment, amend their patches, and push fixes. Currently the GitHub UI does not permit authors to amend commit messages, so this was conceived as a lighter weight workaround.

In addition, this enables targeted remediations. For example, if a branch contains multiple authors, a rebase will add an author's signoff to all commits, even ones they didn't author. This may not be desirable behavior. Both individual and third-party remediations allow an author to retroactively sign off on only the commits they want to fix. This could also be used when doing a bulk remediation on a project which adopted the DCO after being open sourced.

Finally, the goal was to establish standard text that could be replicated and understood by other tools. I specifically avoided an approach that relies upon GitHub tooling (such as signing off in the comment stream or on the PR itself) because this approach embeds the remediation directly in the git log, making it both portable and permanent.

**Due diligence**

I recognize this is a significant addition in functionality. Over the past year I've reviewed the concept with a number of open source legal folks and some of the large projects hosted at the Linux Foundation. The feedback has been generally positive, particularly given that it can be enabled on a project-by-project basis.

**Next steps**

These commits attempt to preserve the existing code style to generate a minimal diff, and make it easier to see what has changed. If merged, it would make sense to follow up with a linted version.

Thanks!

-----
[View rendered README.md](https://github.com/brianwarner/dco/blob/add-remediation-commits/README.md)

### CODE REVIEW

File: test/dco.test.js [hunk 1]
```diff
@@ -1,517 +1,6196 @@
+/*
+* The tests in this file verify the functionality of Probot DCO. Because there are many ways to
+* write a signoff and many ways to misconfigure git, the goal is to flush out as many corner cases
+* as possible.
+*
+* If you are adding a new feature, please note that you might need to repeat your test a few times
+* to verify that functionality is unchanged when various config options are enabled.
+*
+* Usage (from root directory): ./node_modules/jest/bin/jest.js
+* To see a list of tests: ./node_modules/jest/bin/jest.js --verbose
```
Comment (gr2m) [inline test/dco.test.js:10]: I think it's safe to expect people have `npx` now, as it comes with all currently supported node versions. So folks can run `npx jest` or `npx jest --verbose`, which will run the same way on windows.

**Review by gr2m — COMMENTED** [review]
Changes look good to me overall. I'm just concerned that it adds quite some complexity, I am worried about the maintenance of the app. As you use the app at the Linux Foundation, I wonder if you or someone else from your team(s) could help out maintaining DCO, and especially the advanced features added by this PR?

File: index.js [hunk 2]
```diff
@@ -167,10 +168,101 @@ module.exports = ({ app }) => {
...
+    returnMessage = `There are ${dcoFailed.length} commits incorrectly signed off.  This means that the author(s) of these commits failed to include a Signed-off-by line in their commit message.\n\n`
+  }
+
+  returnMessage = returnMessage + `To avoid having PRs blocked in the future, always include \`Signed-off-by: Author Name <authoremail@example.com>\` in *every* commit message. You can also do this automatically by using the -s flag (i.e., \`git commit -s\`).
+
+Here is how to fix the problem so that this code can be merged.\n\n`
+
+  let rebaseWarning = ''
+
+  if (allowRemediationCommits.individual || allowRemediationCommits.thirdparty) {
```
Comment (hiimbex) [inline index.js:186]: When testing the feature, I never made it into this if statement - here's the yaml file i used: https://github.com/hiimbex/Test/blob/master/.github/dco.yaml & the test PR i created: https://github.com/hiimbex/Test/pull/4 - totally could be something wrong in how I'm testing, so let me know, but this scenario was my interpretation from the docs.

File: lib/dco.js [hunk 3]
```diff
@@ -50,36 +52,101 @@ module.exports = async function (commits, isRequiredFor, prURL) {
...
+  if (failed) {
+    return failed.sort((a, b) => a.author.localeCompare(b.author))
+  } else {
+    return []
+  }
 }
 
-function getSignoffs (commit) {
-  const regex = /^Signed-off-by: (.*) <(.*)>$/img
+function getSignoffs (comparisonCommit, comparisonSha, allCommits, allowRemediationCommits) {
```
Comment (hiimbex) [inline lib/dco.js:92]: I find the naming `comparison` a bit confusing - this is referring to the single commit that's currently being processed correct? esp when we get into [this area](https://github.com/probot/dco/pull/147/files#diff-ffe63918db292082603fd2f4cc677c991da9be21a8591c6503d4504efad0b822R115-R118) it gets pretty hard to parse.

**Review by hiimbex — COMMENTED** [review]
So I tested current functionality and feel pretty comfortable merging considering this is an opt in feature, especially with @brianwarner offering to help maintain this feature.

I skimmed the actual logic changes but don't really care to nitpick- would love a follow up styling/refactoring PR - rereading this code hurts my brain 🙈 

I did try to test the new code path, but can't quite get the scenario correct? Left an inline comment where i got stuck.

The tests look fantastic - really appreciate the thoroughness.

File: lib/dco.js (hunk 3 above)
Comment (brianwarner) [reply lib/dco.js:92]: Yes, that's correct and it's a good point. It is confusing. `currentCommit` and `currentSha` might be better? As in, this is the current commit we're processing in an attempt to find valid remediation text in all other commits.

File: index.js (hunk 2 above)
Comment (brianwarner) [reply index.js:186]: Hmm, this is odd. I've been staring at this and must be missing something obvious... it was working earlier in development and appears to have stopped now. It's like it isn't reading the `dco.yml` file correctly and is always defaulting to the `false` values for both. Does anything jump out around line 20?

I guess this wasn't surfaced in the `dco.js` tests because those two variables are defined independent of what's in `dco.yml` for each test. I'm coming up against the limits of my knowledge of jest here, and suspect this would have been detected had there been appropriate snapshot tests that include permutations of `dco.js`. I'll admit I don't know how to do this aside from literally changing the default value in `index.js`, do you have any suggestions?

File: test/dco.test.js (hunk 1 above)
Comment (brianwarner) [reply test/dco.test.js:10]: Thanks! It's a good point, I'll push a commit that fixes this.

**Comment by brianwarner** [comment]
I have to admit, I'm totally stuck on this. For some reason it's no longer reading the config file, and I'm not sure what I'm doing wrong. @hiimbex or @gr2m, it's probably some completely naive mistake on my part... Do you have any suggestions why [this particular line](https://github.com/brianwarner/dco/blob/a914f486d4651fb1b6db431c4fbcaeb0491ac6dd/index.js#L20) would not be reading an optional config file at `.github/dco.yml`?

**Comment by gr2m** [comment]
I don't know out of hand, sorry, I'd have to take a longer look, and I don't have the time right now, sorry. 

Do you see any requests happening if you set `LOG_LEVEL` to `debug`? Also the probot version used in the app is a beta version, it woudl be good to upgrade that to the latest stable version, as there were several fixes since the beta

**Comment by brianwarner** [comment]
Hmm, I upgraded all packages, and now my test instance is crashing on deploy at heroku. I switched back to `master`, verified it deploys properly before running `npm upgrade` (it does), and verified that it crashes after going from 11.0.0-beta3 to 11.1.0 (it does, unfortunately). It doesn't even get to the point of debug logging. The odd thing is that it builds and runs fine locally.

I totally understand the lack of time, and unfortunately it looks like I'm finding more problems than I'm solving. Looks like I have some homework to do.

For what it's worth, here's the error log from heroku:

```
2021-03-16T15:46:49.443667+00:00 heroku[web.1]: State changed from crashed to starting
2021-03-16T15:46:53.281745+00:00 heroku[web.1]: Starting process with command `npm start`
2021-03-16T15:46:56.967922+00:00 app[web.1]:
2021-03-16T15:46:56.967936+00:00 app[web.1]: > dco@1.0.0 start
2021-03-16T15:46:56.967936+00:00 app[web.1]: > probot run ./index.js
2021-03-16T15:46:56.967937+00:00 app[web.1]:
2021-03-16T15:46:58.905343+00:00 app[web.1]: INFO (server): Running Probot v11.1.0 (Node.js: v15.11.0)
2021-03-16T15:46:58.911452+00:00 app[web.1]: INFO (server): Listening on http://localhost:29480
2021-03-16T15:46:58.916887+00:00 app[web.1]: /app/index.js:8
2021-03-16T15:46:58.916888+00:00 app[web.1]: app.on(
2021-03-16T15:46:58.916889+00:00 app[web.1]: ^
2021-03-16T15:46:58.916889+00:00 app[web.1]:
2021-03-16T15:46:58.916890+00:00 app[web.1]: TypeError: Cannot read property 'on' of undefined
2021-03-16T15:46:58.916897+00:00 app[web.1]: at module.exports (/app/index.js:8:7)
2021-03-16T15:46:58.916898+00:00 app[web.1]: at Server.load (/app/node_modules/probot/lib/server/server.js:49:15)
2021-03-16T15:46:58.916898+00:00 app[web.1]: at combinedApps (/app/node_modules/probot/lib/run.js:87:20)
2021-03-16T15:46:58.916899+00:00 app[web.1]: at async Server.load (/app/node_modules/probot/lib/server/server.js:49:9)
2021-03-16T15:46:58.916899+00:00 app[web.1]: at async Object.run (/app/node_modules/probot/lib/run.js:90:9)
2021-03-16T15:46:59.067203+00:00 heroku[web.1]: Process exited with status 1
2021-03-16T15:46:59.157881+00:00 heroku[web.1]: State changed from starting to crashed
```

**Comment by brianwarner** [comment]
Well, I figured one thing out... the config file is read only if it was present in the repo prior to where the PR was forked. So if, for example, you make some changes that result in a dco failure, then update `.github/dco.yml` in the primary branch to enable the new paths, and just close/reopen the original PR, it'll not detect the changes to the config file and looks like it isn't working.

@hiimbex, I was able to use the new functionality by doing the following:
1. Make a change without signoff and open a PR, check that it fails (verifying the default case)
2. Commit a `.github.com/dco.yml` file to the primary branch that enables individual or individual+third party remediations
3. Make a new change without signoff and open a fresh PR, and verify that it provides individual remediations as an option

Taking a step back, I think this makes logical sense. Enabling/disabling remediation commits is a policy change that likely shouldn't be applied retroactively. And if someone opened a PR prior to enabling remediation commits, they should be able to rebase on the main branch.

**Comment by gr2m** [comment]
can we close this PR then? Can we somehow improve the docs to clarify the behavior?

**Comment by brianwarner** [comment]
Hi @gr2m, I'd like to still keep this one open for consideration if possible, please. This feature would be really helpful in some of our projects.

**Comment by bnb** [comment]
@gr2m would you be open to adding CODEOWNERS for just the features being added here?

**Comment by gr2m** [comment]
do you mean to move the new features implemented here to separate files and then use `CODEOWNERS` to enforce reviews / communicate ownership for these files? Sounds good to me

**Comment by gr2m** [comment]
I'm looked at the PR and current state of DCO. I'd like to address a few things before merging to master and deploying

1. Upgrade Probot v11-beta.3 to latest v11 (we can upgrade to v12 later)
2. Increase test coverage. I like to have 100% test coverage, especially for Open Source projects, to accept changes with bigger confidence and speed

For simpler collaboration, I suggest I merge this PR into a local branch, and then start a PR against that branch which addresses the points above.

Once we feel confident I'd do a preview deployment so we can test the changes before deploying them to all users of the DCO app.

File: README.md [hunk 4]
```diff
@@ -10,6 +10,35 @@ a GitHub Integration built with [probot](https://github.com/probot/probot) that
...
+
+Additionally, a project can allow third-parties to sign off on an author's behalf by pushing an additional properly signed-off commit with additional text in the commit log that indicates they sign off on behalf of the author. Third-party remediation requires individual remediation to be enabled.
+
+To enable this, place the following configuration file in `.github/dco.yml` on the default branch:
+
+```yaml
+allowRemediationCommits:
+  individual: true
+  thirdParty: true
+```
```
Comment (gr2m) [inline README.md:40]: Can we add instructions on how to add remediation commits using `git` and the GitHub Web UI? I know you add instructions to the check runs which is great, but I think we should document it here as well, what do you think?

If I understand it correctly, ideally remediation commits would be empty commits with the correct messages, but there is no way to add empty commits using the GitHub Web UI as far as I know.

File: README.md [hunk 5]
```diff
@@ -10,6 +10,35 @@ a GitHub Integration built with [probot](https://github.com/probot/probot) that
...
+### Third-party remediation support
+
+Additionally, a project can allow third-parties to sign off on an author's behalf by pushing an additional properly signed-off commit with additional text in the commit log that indicates they sign off on behalf of the author. Third-party remediation requires individual remediation to be enabled.
+
+To enable this, place the following configuration file in `.github/dco.yml` on the default branch:
+
+```yaml
+allowRemediationCommits:
+  individual: true
+  thirdParty: true
```
Comment (gr2m) [inline README.md:39]: I think setting `thirdParty: true` will set `individual: true` implicitly, setting both is no different to just setting `thirdParty: true`

https://github.com/brianwarner/dco/blob/eb53e5d32ce5d48b1467a581ed0ea404dc0a349c/index.js#L186

If that's the case, how about we change the setting to 

```yml
allowRemediationCommits: individual # or: thirdParty
```

File: index.js (hunk 2 above)
Comment (gr2m) [inline index.js:186]:
```suggestion
  if (allowRemediationCommits.individual || allowRemediationCommits.thirdParty) {
```

Comment (gr2m) [reply index.js:186]: I caught this as I was adding more tests. No need to change the current PR, let's merge this into a local branch (not `master`) then I'll send a PR with my changes against it

File: index.js [hunk 6]
```diff
@@ -167,10 +168,101 @@ module.exports = ({ app }) => {
...
+            returnMessage = returnMessage + '\nSigned-off-by: YOUR_NAME <YOUR_EMAIL>\n```\n'
+          }
+          // Update the currentAuthor and write a new section
+          currentAuthor = commit.author + ' <' + commit.email + '>'
+          returnMessage = returnMessage + `#### On behalf of ${commit.author} &lt;${commit.email}&gt;\n\`\`\`\nThird-Party DCO Remediation Commit for ${commit.author} <${commit.email}>\n\n`
+        }
+        // Draft the magic DCO remediation commit text for the author
+        // returnMessage = returnMessage + `Retroactive-signed-off-by: ${commit.author} <${commit.email}> ${commit.sha}\n`
+        returnMessage = returnMessage + `On behalf of ${commit.author} <${commit.email}>, I, YOUR_NAME <YOUR_EMAIL>, hereby add my Signed-off-by to this commit: ${commit.sha}\n`
+        console.log(commit.sha)
```
Comment (gr2m) [inline index.js:242]: I asume this is a leftover `console.log()`? I'll remove it in my follow up PR
